lib/elixir/lib/module/types/infer.ex

```diff
@@ -102,26 +104,84 @@ defmodule Module.Types.Infer do
 
   # %{...}
   def of_pattern({:%{}, _meta, args} = expr, context) do
-    result =
-      expr_stack(expr, context, fn context ->
-        map_reduce_ok(args, context, fn {key, value}, context ->
-          with {:ok, key_type, context} <- of_pattern(key, context),
-               {:ok, value_type, context} <- of_pattern(value, context),
-               do: {:ok, {key_type, value_type}, context}
-        end)
+    case expr_stack(expr, context, &of_pairs(args, &1)) do
+      {:ok, pairs, context} -> {:ok, {:map, pairs_to_unions(pairs, context)}, context}
+      {:error, reason} -> {:error, reason}
+    end
+  end
+
+  # %var{...}
+  def of_pattern({:%, _meta1, [var, {:%{}, _meta2, args}]} = expr, context) when is_var(var) do
+    expr_stack(expr, context, fn context ->
+      with {:ok, pairs, context} <- of_pairs(args, context),
+           {var_type, context} = new_var(var, context),
+           {:ok, _, context} <- unify(var_type, :atom, context) do
+        pairs = [{{:literal, :__struct__}, var_type} | pairs]
+        {:ok, {:map, pairs_to_unions(pairs, context)}, context}
+      end
+    end)
+  end
+
+  # %Struct{...}
+  def of_pattern({:%, _meta1, [module, {:%{}, _meta2, args}]} = expr, context)
+      when is_atom(module) do
+    struct_pairs =
+      Enum.map(:maps.remove(:__struct__, module.__struct__()), fn {key, value} ->
+        {term_to_type(key), term_to_type(value)}
       end)
 
-    case result do
-      {:ok, pairs, context} -> {:ok, {:map, pairs}, context}
-      {:error, reason} -> {:error, reason}
+    case expr_stack(expr, context, &of_pairs(args, &1)) do
+      {:ok, pattern_pairs, context} ->
+        pairs = pairs_to_unions(pattern_pairs ++ struct_pairs, context)
+        {:ok, {:map, pairs}, context}
+
+      {:error, reason} ->
+        {:error, reason}
     end
   end
 
-  # TODO: structs
-  def of_pattern({:%, _meta, _args}, context) do
-    {:ok, {:map, []}, context}
+  defp of_pairs(pairs, context) do
+    map_reduce_ok(pairs, context, fn {key, value}, context ->
+      with {:ok, key_type, context} <- of_pattern(key, context),
+           {:ok, value_type, context} <- of_pattern(value, context),
+           do: {:ok, {key_type, value_type}, context}
+    end)
+  end
+
+  defp pairs_to_unions(pairs, context) do
+    # Maps only allow simple literal keys in patterns and
+    # term_to_type/1 does not return supertypes so we do
+    # not have to do subtype checking
+
+    Enum.reduce(pairs, [], fn {key, value}, pairs ->
+      case :lists.keyfind(key, 1, pairs) do
+        {^key, {:union, union}} ->
+          :lists.keystore(key, 1, pairs, {key, to_union([value | union], context)})
+
+        {^key, original_value} ->
+          :lists.keystore(key, 1, pairs, {key, to_union([value, original_value], context)})
+
+        false ->
+          [{key, value} | pairs]
+      end
+    end)
   end
 
+  def term_to_type(term) when is_atom(term), do: {:literal, term}
```

Comment:
Let's document all public functions. The fact we have `@moduledoc false` means we won't export it.